pintc/src/predicate/optimize/const_folding.rs

```diff
@@ -30,19 +66,131 @@ pub(crate) fn fold_consts(contract: &mut Contract) {
+/// x * 1 is x
+/// 0 / x is 0
+/// x / 1 is x
+/// x % 1 is x
+pub(crate) fn fold_identities(contract: &mut Contract) -> bool {
+    let mut expr_keys_to_replace: Vec<(ExprKey, ExprKey)> = vec![];
+    let mut did_fold_const: bool = false;
+
+    for pred_key in contract.preds.keys().collect::<Vec<_>>() {
+        for expr_key in contract.exprs(pred_key) {
+            if let Expr::BinaryOp { op, lhs, rhs, .. } = expr_key.get(contract) {
+                let lhs_imm = if let Expr::Immediate { value, .. } = lhs.get(contract) {
+                    Some(value)
+                } else {
+                    None
+                };
+
+                let rhs_imm = if let Expr::Immediate { value, .. } = rhs.get(contract) {
+                    Some(value)
+                } else {
+                    None
+                };
+
+                let replacement_expr_key = match (op, lhs_imm, rhs_imm) {
+                    (BinaryOp::LogicalAnd, Some(Immediate::Bool(true)), _) => Some(*rhs),
+
+                    (BinaryOp::LogicalAnd, _, Some(Immediate::Bool(true))) => Some(*lhs),
+
+                    (BinaryOp::LogicalAnd, Some(Immediate::Bool(false)), _) => Some(*lhs),
+
+                    (BinaryOp::LogicalAnd, _, Some(Immediate::Bool(false))) => Some(*rhs),
+
+                    (BinaryOp::LogicalOr, Some(Immediate::Bool(true)), _) => Some(*lhs),
+
+                    (BinaryOp::LogicalOr, _, Some(Immediate::Bool(true))) => Some(*rhs),
+
+                    (BinaryOp::LogicalOr, Some(Immediate::Bool(false)), _) => Some(*rhs),
+
+                    (BinaryOp::LogicalOr, _, Some(Immediate::Bool(false))) => Some(*lhs),
+
+                    (BinaryOp::Add, Some(Immediate::Int(0)), _) => Some(*rhs),
+
+                    (BinaryOp::Add, _, Some(Immediate::Int(0))) => Some(*lhs),
+
+                    (BinaryOp::Add, Some(Immediate::Real(0.0)), _) => Some(*rhs),
+
+                    (BinaryOp::Add, _, Some(Immediate::Real(0.0))) => Some(*lhs),
+
+                    (BinaryOp::Sub, _, Some(Immediate::Int(0))) => Some(*lhs),
+
+                    (BinaryOp::Sub, _, Some(Immediate::Real(0.0))) => Some(*lhs),
+
+                    (BinaryOp::Mul, Some(Immediate::Int(0)), _) => Some(*lhs),
+
+                    (BinaryOp::Mul, _, Some(Immediate::Int(0))) => Some(*rhs),
+
+                    (BinaryOp::Mul, Some(Immediate::Real(0.0)), _) => Some(*lhs),
+
+                    (BinaryOp::Mul, _, Some(Immediate::Real(0.0))) => Some(*rhs),
+
+                    (BinaryOp::Div, Some(Immediate::Int(0)), _) => Some(*lhs),
+
+                    (BinaryOp::Div, _, Some(Immediate::Int(1))) => Some(*lhs),
+
+                    (BinaryOp::Div, Some(Immediate::Real(0.0)), _) => Some(*lhs),
+
+                    (BinaryOp::Div, _, Some(Immediate::Real(1.0))) => Some(*lhs),
+
+                    (BinaryOp::Mod, _, Some(Immediate::Int(1))) => Some(*lhs),
+
+                    (BinaryOp::Mod, _, Some(Immediate::Real(1.0))) => Some(*lhs),
```

Comment:
Do we allow `Mod` for `real`? If so, that's a bug. The modulo operation should only be supported for integers.

Reply:
Yep, `Mod` is currently allowed for `real`. Will make it an issue and remove it from this PR